viewing_party/party.py

```diff
@@ -1,23 +1,471 @@
+#         {'genre': 'Fantasy', 'rating': 4.8, 'title': 'The Lord of the Functions: The Fellowship of the Function'}... 'rating': 2.0, 'title': 'Recursion'}, 
+#         {'genre': 'Intrigue', 'rating': 4.5, 'title': 'Instructor Student TA Manager'}]}
+
+# def get_watched_avg_rating(user_data):
+#     # This represents that the user has a list of watched movies
+#     watched = user_data.get("watched", [])
+    
+#     # The average rating of an empty watched list is `0.0`
+#     if not watched:
+#         return 0.0
+    
+#     # Calculate the average rating of all movies in the watched list
+#     total_ratings = sum(movie.get("rating", 0.0) for movie in watched)
+    
+#     # return the average rating
+#     avg_rating = total_ratings / len(watched)
+#     return avg_rating
+
+# ========================================= wave 02- # 2. get_most_watched_genre ========== RC
+
+# def get_most_watched_genre(user_data):
+#     watched_genres = {}
+
+# # - Determine which genre is most frequently occurring in the watched list
+#     # iterate through 
+#     for movie in user_data["watched"]:
+#         genre = movie["genre"]
+#         if genre in watched_genres:
+#             watched_genres["genre"] += 1
+#         else:
+#             watched_genres["genre"] = 1
+
+#     if not watched_genres:
+#         return None 
+#     print(watched_genres)
+#     return max(watched_genres, key=watched_genres.get)
+
+
+# ========================================= wave 02- # 2. get_watched_avg_rating ========== SJ
+
+def get_watched_avg_rating(user_data): 
+    
+    #  initialize a variable for ratings  set to 0     
+    ratings_total = 0
+    average_rating = 0
+
+    #  if the length of the user_ data( "watched") is empty  
+    if len(user_data["watched"]) == 0: 
+        average_rating == 0.0
+        return average_rating
+    
+    else: 
+        # loop through movies in the the watched list in user_data
+        for movie in user_data["watched"]: 
+                            
+            #  add the rating to the ratings total 
+            ratings_total += movie["rating"]
+            print(ratings_total)
+            
+            #  average rating is ratings_total / length (user_data["watched"]) 
+            average_rating = ratings_total/ len(user_data["watched"])
+                
+    return average_rating
+
+    
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+# ========================================= wave 03- # 1. get_unique_watched ========== RC
+
+def get_unique_watched(user_data):
+    # movies that friends have watched
+    friends_watched = {movie['title'] for friend in user_data['friends'] for movie in friend['watched']}
+
+    # movies that ONLY user has watched 
+    unique_movies = [movie for movie in user_data['watched'] if movie['title'] not in friends_watched]
+
+    # Return the list of unique movies
+    return unique_movies

```

Comment:
Amazing work with this dictionary comprehension, just be sure you can understand it in the future 💯